scripts/factory-common-install/libexec/app_current_device_vnc.sh

```diff
@@ -0,0 +1,167 @@
+    fi
+
+    if is_x11vnc_pid_stopped "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${vnc_remote_port}"; then
+        echo "x11vnc pid is already stopped"
+    else
+        echo "Closing x11vnc pid"
+        exit_x11vnc "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}"
+    fi
+
+    return 0
+  }
+  trap "cleanup '${vnc_user}' '${vnc_hostname}' '${device_ssh_port}' '${vnc_remote_port}' '${systemd_service_name}'" EXIT SIGINT SIGQUIT
+
+  user_info="$(get_factory_info__user_full_name)<$(get_factory_info__user_email)>"
+  vnc_passwd="$(pwgen -cn 32 | head -n1)"
+
+  # Validation that the service that we are about to launch does not exist yet.
+  # if the service is alive we ask the user if it wants to kill the existing session.
+  echo " -> Making sure x11vnc is not aready started"
+  echo '"${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"'
+  if ! is_x11vnc_systemd_process_stopped "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"; then
+    active_session_info=$(ssh "${vnc_user}@${vnc_hostname}" -p "${device_ssh_port}" "systemctl -q --no-pager --user status '${systemd_service_name}'") && true
+    echo "An active x11vnc session has been detected"
+    echo "Here is some information about the active session"
+    echo "${active_session_info}"
+    echo ""
+    read -n 1 -p 'Do you want to close the session? (Y)es or (N)o' user_input
+    echo ""
+    case $user_input in
+      [yY] )
+        echo "Okay, we will kill the existing session"
+        cleanup "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${vnc_remote_port}" "${systemd_service_name}"
+        # Sleep 5 seconds to avoid getting kicked out by the other script cleanup.
+        sleep 5
+        ;;
+      [nN] )
+        echo "Okay, leaving the existing session as is."
+        trap - EXIT SIGINT SIGQUIT
+        return 0
+        ;;
+      * )
+        echo "Erroneous answer. Doing nothing."
+        trap - EXIT SIGINT SIGQUIT
+        return 0
+        ;;
+    esac
+  fi
+
+  # x11vnc is launched as a connect once server, as soon as vncviewer closes x11vnc also closes.
+  echo " -> Starting x11vnc"
+  cmd="$(cat <<EOF
+systemd-run \
+--remain-after-exit \
+--property=Type=forking \
+--unit "${systemd_service_name}" \
+--user bash -c 'echo "${user_info}"'; x11vnc -bg -once -localhost -passwd "${vnc_passwd}" -display :0
+EOF
+)"
+  if ! ssh "${vnc_user}@${vnc_hostname}" -p "${device_ssh_port}" "${cmd}"; then
+    err="$?"
+    1>&2 echo "Error: $FUNCNAME: $err"
+    return "$err"
+  fi
+
+  # The tunnel auto-closes as soon as vncviewer is closed.
+  # Nothing to clean after.
+  echo " -> Creating an SSH tunnel"
+  if ! ssh -f -T -L "${vnc_local_port}:localhost:${vnc_remote_port}" "${vnc_user}@${vnc_hostname}" -p "${device_ssh_port}" "sleep 10"; echo "${vnc_passwd}" | vncviewer -autopass localhost:"${vnc_local_port}"; then
```

Comment:
This 10s sleep seems arbitrary. Any reason why we do this? Can't we instead detect the condition required to launch the viewer?

Reply:
It is indeed somewhat arbitary. The sleep allows us to have an auto-closing ssh tunnel.
It keeps the tunnel open for 10 sec, during that time we launch the viewer. As soon as the viewer closes, the tunnel also closes.

We could probably set a lower value, though.

Reply:
Ok, as discussed, I now understand that this `;` syntax is not mere bash sequencing but something specific to ssh tunneling. What I got from our discussion is that both the sleep and the vncviewer command will be started in parellel and the tunnel will stay open for the longest of the two command / as long as one of these commands is still alive.